src/hotspot/os/linux/cgroupSubsystem_linux.hpp

```diff
@@ -69,148 +69,99 @@
+{                                                                                     \
+  bool is_ok;                                                                         \
+  is_ok = controller->read_number_handle_max(filename, &retval);                      \
+  if (!is_ok) {                                                                       \
+    log_trace(os, container)(log_string " failed: %d", OSCONTAINER_ERROR);            \
+    return OSCONTAINER_ERROR;                                                         \
+  }                                                                                   \
+  log_trace(os, container)(log_string " is: " JLONG_FORMAT, retval);                  \
 }
 
-#define GET_CONTAINER_INFO_LINE(return_type, controller, filename,        \
-                           matchline, logstring, scan_fmt, variable)      \
-  return_type variable;                                                   \
-{                                                                         \
-  int err;                                                                \
-  err = subsystem_file_line_contents(controller,                          \
-                                filename,                                 \
-                                matchline,                                \
-                                scan_fmt,                                 \
-                                &variable);                               \
-  if (err != 0)                                                           \
-    return (return_type) OSCONTAINER_ERROR;                               \
-                                                                          \
-  log_trace(os, container)(logstring, variable);                          \
+#define CONTAINER_READ_STRING_CHECKED(controller, filename, log_string, retval)       \
+{                                                                                     \
+  bool is_ok;                                                                         \
+  is_ok = controller->read_string(filename, retval);                                  \
+  if (!is_ok) {                                                                       \
+    log_trace(os, container)(log_string " failed: %d", OSCONTAINER_ERROR);            \
+    return nullptr;                                                                   \
+  }                                                                                   \
+  log_trace(os, container)(log_string " is: %s", retval);                             \
 }
 
+/*
+ * Used to model the tuple-valued cgroup interface files like cpu.max, which
+ * contains two values: <quota> <period>. In this case the first tuple value
+ * is <quota> and the second tuple value is <period>. We use this to map the
+ * tuple value to a constant string format for sscanf reading.
+ */
+enum class TupleValue { FIRST, SECOND };
```

Comment:
Style, nit: No reason for all-caps for these names, can just be "First", "Second".